File: TextShooter/BIDViewController.m
//
//  BIDViewController.m
//  TextShooter
//
//  Created by JN on 2013-12-6.
//  Copyright (c) 2013 Apress. All rights reserved.
//

#import "BIDViewController.h"
#import "BIDLevelScene.h"
#import "BIDStartScene.h"
#import "RBSGameCenterManager.h"

@implementation BIDViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    // Configure the view.
    SKView * skView = (SKView *)self.view;
//    skView.showsFPS = YES;
//    skView.showsNodeCount = YES;
    
    // Create and configure the scene.
    SKScene * scene = [BIDStartScene sceneWithSize:skView.bounds.size];
    scene.scaleMode = SKSceneScaleModeAspectFill;
    
    [self authenticateLocalPlayer];
    
    // Present the scene.
    [skView presentScene:scene];
}

- (BOOL)shouldAutorotate
{
    return YES;
}

- (NSUInteger)supportedInterfaceOrientations
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        return UIInterfaceOrientationMaskAllButUpsideDown;
    } else {
        return UIInterfaceOrientationMaskAll;
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Release any cached data, images, etc that aren't in use.
}

- (BOOL)prefersStatusBarHidden {
    return YES;
}

- (void)authenticateLocalPlayer {
    RBSGameCenterManager *gcm =[RBSGameCenterManager shared];
    gcm.shouldShowAuthenticationDialogImmediately = ^BOOL(void) {
        SKView *skView = (SKView *)self.view;
        return [skView.scene isKindOfClass:[BIDStartScene class]];
    };
    [gcm authenticateLocalPlayer:^(BOOL success) {
        SKView *skView = (SKView *)self.view;
        BIDStartScene *startScene = [skView.scene isKindOfClass:[BIDStartScene class]] ? (id)skView.scene : nil;
        if (success) {
//            NSLog(@"playerAuthenticated %@", [GKLocalPlayer localPlayer]);
            startScene.gameCenterButtonEnabled = YES;
        } else {
//            NSLog(@"disableGameCenter");
            startScene.gameCenterButtonEnabled = NO;
        }
    }];
}

@end

File: TextShooter/BIDViewController.swift
//
//  BIDViewController.swift
//  TextShooter
//
//  Created by Jack Nutting on 2025-08-04.
//  Copyright © 2025 Apress. All rights reserved.
//

import AppStoreCatalog
import SpriteKit
import SwiftUI
import UIKit

class BIDViewController: UIViewController {
    private var catalog: AppStoreCatalog?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        catalog = AppStoreCatalog.embeddedCatalog
        Task {
            do {
                catalog = try await AppStoreCatalog.remoteCatalog()
            } catch {
                print("Fetching remote catalog failed: \(error)")
            }
        }
        
        let skView = self.view as! SKView
        
        let scene: SKScene = BIDStartScene(size: skView.bounds.size)
        scene.scaleMode = .aspectFill
        
        authenticateLocalPlayer()
        
        skView.presentScene(scene)
    }
    
    override var shouldAutorotate: Bool { return true }
    override var prefersStatusBarHidden: Bool { return true }
    
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        if UIDevice.current.userInterfaceIdiom == .pad {
            return .all
        } else {
            return .allButUpsideDown
        }
    }
    
    private func authenticateLocalPlayer() {
        let gcm = RBSGameCenterManager.shared()!
        gcm.shouldShowAuthenticationDialogImmediately = {
            let skView = self.view as! SKView
            return skView.scene!.isKind(of: BIDStartScene.self)
        }
        
        gcm.authenticateLocalPlayer { success in
            let skView = self.view as! SKView
            let startScene = skView.scene!.isKind(of: BIDStartScene.self) ? skView.scene as? BIDStartScene : nil
            if success {
                startScene?.gameCenterButtonEnabled = true
            } else {
                startScene?.gameCenterButtonEnabled = false
            }
        }
    }
    
    @objc func showOtherApps() {
        guard let catalog else { return }
        let hostingController = UIHostingController(rootView: AppStoreCatalogView(catalog: catalog, enableCloseButton: true))
        
        present(hostingController, animated: true)
    }
}

extension AppStoreCatalog {
    // This uses force-unwrap and force-try. Valid in this case, as the resources are required to be present in the build, and valid. If any of these fail due to changes, this will crash early in development builds.
    static var embeddedCatalog: AppStoreCatalog {
        let catalogUrl = Bundle.main.url(forResource: "RebisoftAppCatalog", withExtension: "json")!
        let data = try! Data(contentsOf: catalogUrl)
        let catalog = try! AppStoreCatalog(data: data, excluding: "814735204")
        return catalog
    }
    
    static func remoteCatalog() async throws -> AppStoreCatalog {
        let url = URL(string: "https://rebisoft.com/xp/RebisoftAppCatalog.json")!
        let (data, _) = try await URLSession.shared.data(from: url)

        return try AppStoreCatalog(data: data, excluding: "814735204")
    }
}
